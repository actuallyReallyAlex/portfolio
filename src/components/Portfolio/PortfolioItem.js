import React from 'react'
import PropTypes from 'prop-types'
import { Anchor, Box, Heading, Text } from 'grommet'
import { Code, LinkNext } from 'grommet-icons'
import Tilt from 'react-tilt'

const PortfolioItem = ({
  category,
  description,
  image,
  languages,
  link,
  size,
  source,
  title
}) => {
  return (
    <Tilt
      className="Tilt"
      options={{ max: 15, perspective: 2000, scale: 1.05 }}
    >
      <Box
        background="brand"
        className="Tilt-inner"
        width={size === 'small' ? '300' : '384px'}
      >
        <Box
          background={`url(${image})`}
          height="189px"
          width={size === 'small' ? '300' : '384px'}
        />
        <Box pad="medium">
          <Text size="medium">{category}</Text>
          <Text size="xlarge" weight="bold">
            {title}
          </Text>
          <Box margin={{ vertical: 'medium' }}>
            {languages.map(({ color, title }, i) => (
              <Box
                align="center"
                direction="row"
                gap="small"
                key={i}
                justify="start"
              >
                <Box background={color} height="10px" width="10px" />
                <Heading level="4" margin="none">
                  {title}
                </Heading>
              </Box>
            ))}
          </Box>
          <Box fill="horizontal">
            <Text>{description}</Text>
          </Box>
          <Box margin={{ top: 'medium' }}>
            {source && (
              <Box direction="row" gap="small">
                <Anchor
                  color="white"
                  href={source.href}
                  icon={<Code />}
                  label={source.title}
                  rel="noreferrer"
                  target="_blank"
                />
              </Box>
            )}
            {link && (
              <Box direction="row" gap="small">
                <Anchor
                  color="white"
                  href={link.href}
                  icon={<LinkNext />}
                  label={link.title}
                  rel="noreferrer"
                  target="_blank"
                />
              </Box>
            )}
          </Box>
        </Box>
      </Box>
    </Tilt>
  )
}

PortfolioItem.propTypes = {
  category: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  languages: PropTypes.array.isRequired,
  link: PropTypes.object,
  size: PropTypes.string.isRequired,
  source: PropTypes.object,
  title: PropTypes.string.isRequired
}

export default PortfolioItem
