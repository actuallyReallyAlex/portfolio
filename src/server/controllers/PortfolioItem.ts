import express, { Express, Request, Response, Router } from "express";
import multer, { FileFilterCallback } from "multer";
import path from "path";
import { v4 as uuidv4 } from "uuid";

import PortfolioItemModel from "../models/PortfolioItem";

import {
  ErrorResponse,
  PortfolioItem,
  PortfolioItemDocument,
  PortfolioItemModifyResponse,
} from "../types";

class PortfolioItemController {
  public router: Router = express.Router();

  private parseFile = multer({
    fileFilter: (
      req: express.Request,
      file: Express.Multer.File,
      cb: FileFilterCallback
    ) => {
      console.log(`------ parsefile ------`);
      if (
        !file.originalname.match(/\.(png)$/) &&
        !file.originalname.match(/\.(jpg)$/) &&
        !file.originalname.match(/\.(jpeg)$/)
      ) {
        return cb(
          new Error(
            "File must be in one of the following formats: [.png, .jpg, .jpeg]."
          )
        );
      }

      return cb(null, true);
    },
    limits: {
      fileSize: 5000000,
    },
    storage: multer.diskStorage({
      destination: function (req, file, cb) {
        console.log("------ destination ------");
        console.log(path.join(__dirname, "../../uploads"));
        cb(null, path.join(__dirname, "../../uploads"));
      },
      filename: function (req, file, cb) {
        console.log("------ filename ------");
        console.log(`${uuidv4()}.${file.originalname.split(".")[1]}`);
        cb(null, `${uuidv4()}.${file.originalname.split(".")[1]}`);
      },
    }),
  });

  constructor() {
    this.initializeRoutes();
  }

  public initializeRoutes(): void {
    this.router.get(
      "/portfolioItems",
      async (
        req: Request,
        res: Response
      ): Promise<Response<ErrorResponse | PortfolioItemDocument[]>> => {
        try {
          const portfolioItemDocuments = await PortfolioItemModel.find({});

          return res.send(portfolioItemDocuments);
        } catch (error) {
          console.error(error);
          return res.status(500).send({ error });
        }
      }
    );

    this.router.post(
      "/portfolioItem",
      // this.parseFile.single("file"),
      async (req: Request, res: Response): Promise<any> => {
        console.log("------ callback ------");

        const uploadTest = multer({
          fileFilter: (
            req: express.Request,
            file: Express.Multer.File,
            cb: FileFilterCallback
          ) => {
            console.log("------ parseFile ------");
            if (
              !file.originalname.match(/\.(png)$/) &&
              !file.originalname.match(/\.(jpg)$/) &&
              !file.originalname.match(/\.(jpeg)$/)
            ) {
              return cb(
                new Error(
                  "File must be in one of the following formats: [.png, .jpg, .jpeg]."
                )
              );
            }

            return cb(null, true);
          },
          limits: {
            fileSize: 5000000,
          },
          storage: multer.diskStorage({
            destination: function (req, file, cb) {
              console.log("------ destination ------");
              console.log(path.join(__dirname, "../../uploads"));
              cb(null, path.join(__dirname, "../../uploads"));
            },
            filename: function (req, file, cb) {
              console.log("------ filename ------");
              const filename = `${uuidv4()}.${file.originalname.split(".")[1]}`;
              console.log({ filename });
              cb(null, filename);
            },
          }),
        }).single("file");

        uploadTest(req, res, async (err: any) => {
          if (err instanceof multer.MulterError) {
            // A Multer error occurred when uploading.
            console.log("A multer error occured when uploading");
            console.error(err);
            return res.status(500).send({ error: err });
          } else if (err) {
            // An unknown error occurred when uploading.
            console.log("An unknown error occured when uploading");
            console.error(err);
            return res.status(500).send({ error: err });
          }

          // Everything went fine.
          console.log("------ everything is fine ------");

          try {
            const {
              content,
              iconBackground,
              iconClass,
              links,
              tagline,
              title,
            } = req.body;
            console.log({ body: JSON.stringify(req.body, null, 2) });
            const newItemData: PortfolioItem = {
              content,
              coverImage: `/uploads/${req.file.filename}`,
              iconBackground,
              iconClass,
              links: JSON.parse(links),
              tagline,
              title,
            };
            console.log({ newItemData });
            const newPortfolioItem = new PortfolioItemModel(newItemData);
            console.log({ newPortfolioItem });

            await newPortfolioItem.save();

            return res.status(201).send(newPortfolioItem);
          } catch (error) {
            console.error(error);
            return res.status(500).send({ error });
          }
        });
      }
    );

    // TODO - Implement CoverImage modification
    this.router.patch(
      "/portfolioItem",
      this.parseFile.single("file"),
      async (
        req: Request,
        res: Response
      ): Promise<Response<ErrorResponse | PortfolioItemModifyResponse>> => {
        try {
          const correspondingPortfolioItem = await PortfolioItemModel.findById(
            req.body.id
          );
          if (!correspondingPortfolioItem) {
            return res
              .status(404)
              .send({ error: "No corresponding PortfolioItem found!" });
          }

          const updateFieldNames = Object.keys(req.body);

          updateFieldNames.forEach((field) => {
            if (req.body[field] !== correspondingPortfolioItem.get(field)) {
              correspondingPortfolioItem.set(field, req.body[field]);
            }
          });

          await correspondingPortfolioItem.save();

          const portfolioItems = await PortfolioItemModel.find({});

          return res.send({
            portfolioItem: correspondingPortfolioItem,
            portfolioItems,
          });
        } catch (error) {
          console.error(error);
          return res.status(500).send({ error });
        }
      }
    );

    this.router.delete(
      "/portfolioItem",
      async (
        req: Request,
        res: Response
      ): Promise<Response<ErrorResponse | PortfolioItemModifyResponse>> => {
        try {
          const { id } = req.body;

          const item = await PortfolioItemModel.findOneAndDelete({ _id: id });

          if (!item) {
            return res
              .status(404)
              .send({ error: "No corresponding PortfolioItem found!" });
          }

          const portfolioItems = await PortfolioItemModel.find({});

          return res.send({
            portfolioItem: item,
            portfolioItems,
          });
        } catch (error) {
          console.error(error);
          return res.status(500).send({ error });
        }
      }
    );
  }
}

export default PortfolioItemController;
