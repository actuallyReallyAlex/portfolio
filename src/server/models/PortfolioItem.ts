import mongoose from "mongoose";

import { PortfolioItemDocument } from "../types";

const portfolioItemSchema = new mongoose.Schema(
  {
    content: {
      required: [true, "Field 'content' is required."],
      type: String,
    },
    coverImage: {
      required: [true, "Field 'coverImage' is required."],
      type: String,
    },
    iconBackground: {
      required: [true, "Field 'iconBackground' is required."],
      type: String,
    },
    iconClass: {
      required: [true, "Field 'iconClass' is required."],
      type: String,
    },
    links: {
      demo: {
        required: false,
        type: String,
      },
      github: {
        required: false,
        type: String,
      },
      npm: {
        required: false,
        type: String,
      },
    },
    tagline: {
      required: [true, "Field 'tagline' is required."],
      type: String,
    },
    title: {
      required: [true, "Field 'title' is required."],
      type: String,
    },
  },
  { timestamps: true }
);

const PortfolioItem = mongoose.model<PortfolioItemDocument>(
  "PortfolioItem",
  portfolioItemSchema
);

export default PortfolioItem;
