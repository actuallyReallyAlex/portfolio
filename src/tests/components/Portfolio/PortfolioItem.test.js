import React from 'react'
import PortfolioItem from '../../../components/Portfolio/PortfolioItem'
import { shallow } from 'enzyme'

const defaultProps = {
  category: 'Category',
  description: 'Description',
  image: 'fakeURL',
  languages: [{ color: 'accent-1', title: 'JavaScript' }],
  title: "Title"
}

describe('<PortfolioItem />', () => {
  test('Should render the <PortfolioItem />.', () => {
    const component = shallow(<PortfolioItem {...defaultProps} />)
    expect(component).toMatchSnapshot()
  })
})
