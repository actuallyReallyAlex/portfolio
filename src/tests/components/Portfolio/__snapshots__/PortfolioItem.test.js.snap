// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PortfolioItem /> Should render the <PortfolioItem />. 1`] = `ShallowWrapper {}`;
